### PR TITLE
Code improvement:

### PR DESCRIPTION
**Pull request by blasvicco**
I Suggest to use a recursive function instead of cascade of if

### CODE REVIEW

**Comment by jedmccaleb** [comment]
Hi blasvicco, thanks for the PR. Before we can merge it can you agree to the contributor license:  https://docs.google.com/forms/d/1g7EF6PERciwn7zfmfke5Sir2n10yddGGSXyZsq98tVY/viewform

**Comment by blasvicco** [comment]
Hi jedmccaleb, agreement signed!

Regards

File: src/bucket/BucketManagerImpl.cpp [hunk 1]
```diff
@@ -295,28 +295,17 @@ BucketManagerImpl::snapshotLedger(LedgerHeader& currentHeader)
...
-
-                    currentHeader.skipList[3] = currentHeader.skipList[2];
-                }
-                currentHeader.skipList[2] = currentHeader.skipList[1];
-            }
-            currentHeader.skipList[1] = currentHeader.skipList[0];
-        }
-        currentHeader.skipList[0] = currentHeader.bucketListHash;
+        unsigned int next = index + 1;
+        calculateSkipValues(currentHeader, next, skipd + skipV[index]);
```
Comment (vogel) [inline src/bucket/BucketManagerImpl.cpp]: I think it lacks checking if next <= 3

Comment (blasvicco) [inline src/bucket/BucketManagerImpl.cpp]: Yeap, you're right =)
What do you think about something like this:

`
if (skipV.size() > next && currentHeader.skipList.size() > next) {
     calculateSkipValues(currentHeader, next, skipd + skipV[index]);
}
`

Comment (vogel) [inline src/bucket/BucketManagerImpl.cpp]: Thanks. I'll look at this soon.

**Comment by vogel** [comment]
Sorry it took so long. I think that patch can get it when it passes all our tests (I'll run them as soon as previous rounds stops), but there is one more change that I would like to see.

Values of SKIP_1..SKIP_4 and skipV are only used in BucketManagerImpl.cpp file. So it makes sense to put them as a free values in anonymous namespace inside of BucketManagerImpl.cpp. Something like:

namespace
{

auto const SKIP_1 = 50;
auto const SKIP_V = std::vector{....}
}

Usage of SKIP_1 and other in BucketTests.cpp should be changed to just numeric constants (ie: replace header.ledgerSeq = SKIP_3 + SKIP_2; with header.ledgerSeq = 55000;) - because currently one can change values of these constants and do not notice it is wrong until a stellar-core with these values fails to catchup to network.

Thanks again for the contribution.

**Comment by vogel** [comment]
Hello.

Thanks for this, but I think we should go with iterative solution. Recursion makes this code very hard to follow.

I'm thinking about something similar to this:

`
void
BucketManagerImpl::calculateSkipValues(LedgerHeader& currentHeader)
{
    auto skips = std::vector{50, 5000, 50000, 500000};
    auto sum = std::accumulate(std::begin(skips), std::end(skips));
    
    auto v = currentHeader.ledgerSeq - sum;
    for (auto i = skips.length() - 1; i > 1; i++)
    {
        sum -= skips[i];
        auto v = currentHeader.ledgerSeq - sum;
        if (v - sum > 0 && (v % skips[i] == 0))
        {
            currentHeader.skipList[i] = currentHeader.skipList[i - 1];
        }
    }
    
    auto v = currentHeader.ledgerSeq;
    if (v > 0 && v & (v % skips[0] == 0))
    {
        currentHeader.skipList[0] = currentHeader.bucketListHash;
    }
}`

What do you think?

**Comment by blasvicco** [comment]
Hi.
Sorry for the delay. I'm currently in New Zealand because I was one of the selected candidate for the LookSee programe and I'm a little bit busy right now.
I think that remove the #define for each SKIP value is a great idea.
Your suggestion about it:
`
  auto skips = std::vector{50, 5000, 50000, 500000};
  auto sum = std::accumulate(std::begin(skips), std::end(skips));
`
Looks great for me!!!

Some times the recursion are difficult to follow. I'm agree. If this is the case then the iterative alternative could be the solution but, in the suggestion above there are a couple of thing I'm worry about. I couldn't test it but the for doesn't look well for me (maybe I'm wrong):
`for (auto i = skips.length() - 1; i > 1; i++)`
should be
`for (auto i = skips.length() - 1; i > 1; i--)`
to avoid infinite loop, but then we are changing the order of the iteration on the skip vector, maybe that's not a problem but if it is a problem then we can change the for to make it looks like this
`
  unsigned int len = skips.length();
  for (auto i = 0; i < len; i++)
`

On the other hand we have this piece of code:
`
  auto v = currentHeader.ledgerSeq;
  if (v > 0 && v & (v % skips[0] == 0))
  {
      currentHeader.skipList[0] = currentHeader.bucketListHash;
  }
`
That looks very similar to the piece of code inside the for.
I believe it will be great if we can eliminate that somehow. I currently have not much time, but maybe next week I can give a better look to the iterative solution and avoid the code duplication.
What do you think about it?

Thanks

**Comment by vogel** [comment]
Hi.

You are right about the loop, this is not the final implementation, just the idea. What about this:

```
void
BucketManagerImpl::calculateSkipValues(LedgerHeader& currentHeader)
{
auto skips = std::vector{50, 5000, 50000, 500000};
auto sum = std::accumulate(std::begin(skips), std::end(skips));
auto replacements = std::vector<...>{&currentHeader.bucketListHash, &currentHeader.skipList[0], &currentHeader.skipList[1], &currentHeader.skipList[2]};

  auto v = currentHeader.ledgerSeq - sum;
  for (auto i = skips.length() - 1; i > 0; i--)
  {
      sum -= skips[i];
      auto v = currentHeader.ledgerSeq - sum;
      if (v - sum > 0 && (v % skips[i] == 0))
      {
          currentHeader.skipList[i] = replacements[i];
      }
  }
}
```

**Comment by blasvicco** [comment]
Awesome Rafal!!!

**Comment by vogel** [comment]
Thanks again. Replaced by #1222
